scenes/cslib_object.txt

```diff
@@ -110,3 +159,19 @@
 	*set n + 1
 	*goto _set_main_loop
 *return
+
+*comment --- ------------------------------------------------------------------
+*comment --- Sets all the fields of the main instance to false
+*comment --- ------------------------------------------------------------------
+*label _clear_main_instance
+*params p_object
+*temp n 1
+*temp field_count {p_object&"_field_max"}
+*temp field ""
+*label _clear_main_loop
+*if (n <= field_count)
+	*set field {p_object&"_field_${n}"}
+	*set {p_object&"_${field}"} false
```

Comment:
I think we should use "" rather than _false_ here, to keep it consistent with resetting cslib_ret. It's not perfect, but it's the closest we have to null/undefined. _false_ is more likely to clash with an intended value (e.g. dragon_alive: false).

Reply:
I think I'd prefer the crash: explicit errors are much easier to diagnose than logical ones. What do you think @diegoboston ?

Reply:
And @danspinola ?

Reply:
I don't think using a library should cause (perfectly legal) code to fail.

It might put people off using it apart from anything else, and that would be a waste. :)

Reply:
It's pretty much an intractable problem.

Unless we require people to define field types to which I say 'no - what on earth are you thinking?!

We could, in theory, provide a configuration option but that would involve another cslib global which wouldn't be a good idea either.

I'm still in favour of _false_, but I'll go with the majority..

Forcing an _id field is kind of a separate issue although I'm not against it. Setting it to a negative (or whatever) value is useful but you still have the question of what to do with the rest of the fields - do you clear them or leave them with their current (wrong) values?

Reply:
This may ultimately not matter all that much.

I find that developing with ChoiceScript basically involves stopping it crashing. If I fail to find dragon 'Fnord' and it barfs I go back and look at it until I realise that I made a typo and I meant 'Ford'.

Checking return values is seldom worth the effort. (I know, I know. Bad programming style and all that but it seems to be what happens)

Reply:
Yet another alternative would be to use Unix-style error codes and messages.

Required fields:
*create dragon_err
*create dragon_errmsg

Where dragon_err will be 0 if there is no error and dragon_errmsg is optional.

This might be overkill.

Reply:
It sounds to me like you might be in favour of throwing a *bug instead of clearing on a bad search value? Shall we go with that? As you say, it is quite ChoiceScriptish. But, more importantly, it means we don't have to come to any more controversial conclusion of what to do with the fields. We can punt this problem down the road.

Reply:
Works for me. At least it's equally bad for everyone...